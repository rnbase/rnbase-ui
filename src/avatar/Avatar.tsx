import React, { useCallback, useEffect, useMemo, useState } from 'react';
import md5 from 'js-md5';
import {
  Image,
  ImageStyle,
  ImageSourcePropType,
  LayoutChangeEvent,
  PixelRatio,
  StyleSheet,
  StyleProp,
  Text,
  TextStyle,
  View,
  ViewProps,
  ViewStyle,
} from 'react-native';

import { Themed, Theme, WithThemeProps, withTheme } from '../theming';
import { getColor, getInitials, getRadius } from '../helpers';
import Badge, { Props as BadgeProps } from '../badge/Badge';

interface AvatarProps extends ViewProps {
  size?: number;
  cornerRadius?: number | string;
  name?: string;
  email?: string;
  colorize?: boolean;
  badge?: BadgeProps;
  badgePosition?: 'top-right' | 'bottom-right' | 'bottom-left' | 'top-left';
  imageSource?: ImageSourcePropType;
  defaultImageSource?: ImageSourcePropType;
  style?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  imageStyle?: StyleProp<ImageStyle>;
}

const Avatar: React.FC<Themed<typeof createStyleSheet, AvatarProps>> = ({
  theme: { styles },
  size = 50,
  cornerRadius = '50%',
  name,
  email,
  colorize = false,
  badge,
  badgePosition = 'top-right',
  imageSource,
  defaultImageSource = require('./default.png'),
  style,
  textStyle,
  imageStyle,
  ...props
}) => {
  const getAvatarImageSource = useCallback(() => {
    if (imageSource) {
      return imageSource;
    } else if (email) {
      const emailHash = md5(email.toLowerCase().trim());
      const pixelSize = PixelRatio.getPixelSizeForLayoutSize(size);

      return { uri: `https://www.gravatar.com/avatar/${emailHash}?s=${pixelSize}&d=404` };
    }

    return defaultImageSource;
  }, [imageSource, email, size, defaultImageSource]);

  const [badgeHeight, setBadgeHeight] = useState(10);
  const [imageHeight, setImageHeight] = useState(size);
  const [avatarImageSource, setAvatarImageSource] = useState(getAvatarImageSource);

  useEffect(() => setAvatarImageSource(getAvatarImageSource()), [
    getAvatarImageSource,
    setAvatarImageSource,
  ]);

  const onError = useCallback(() => setAvatarImageSource(defaultImageSource), [defaultImageSource]);

  const onImageLayout = useCallback((event: LayoutChangeEvent) => {
    setImageHeight(event.nativeEvent.layout.height);
  }, []);

  const onBadgeLayout = useCallback((event: LayoutChangeEvent) => {
    setBadgeHeight(event.nativeEvent.layout.height);
  }, []);

  const initials = useMemo(
    () =>
      name &&
      avatarImageSource === defaultImageSource && {
        color: colorize ? getColor(name) : undefined,
        text: getInitials(name),
      },
    [name, colorize, avatarImageSource, defaultImageSource],
  );

  const borderRadius = getRadius(cornerRadius, size);

  const badgeOffset = useMemo(() => {
    // We want to place the badge at the point with polar coordinates (r,45°)
    // thus we need to find the distance from the containing square top right corner
    // which can be calculated as x = r - r × sin(45°)
    // Self offset is how much we’ll shift the badge from the edge point,
    // its value ranges from badgeHeight / 4 (square) to badgeHeight / 2 (circle)
    const selfOffset = badgeHeight * (0.25 + borderRadius / (size * 2));
    const edgeOffset = borderRadius * (1 - Math.sin((45 * Math.PI) / 180));

    return PixelRatio.roundToNearestPixel(edgeOffset - selfOffset);
  }, [badgeHeight, borderRadius, size]);

  const rootStyles = [
    {
      width: size,
      height: size,
      borderRadius,
    },
    styles.root,
    style,
  ];

  let content = [];

  if (initials) {
    const fontSize = PixelRatio.roundToNearestPixel(size / 2.5);

    content.push(
      <Text key="text" style={[{ fontSize }, styles.text, textStyle]} numberOfLines={1}>
        {initials.text}
      </Text>,
    );

    if (initials.color) {
      rootStyles.push({ backgroundColor: initials.color });
    }
  } else {
    const imageRadius = borderRadius - (size - imageHeight) / 2;

    content.push(
      <Image
        key="image"
        style={[{ borderRadius: imageRadius }, styles.image, imageStyle]}
        source={avatarImageSource}
        onLayout={onImageLayout}
        onError={onError}
      />,
    );
  }

  if (badge) {
    const [badgeY, badgeX] = badgePosition.split('-');
    const badgeStyles = [
      {
        [badgeY]: badgeOffset,
        [badgeX]: badgeOffset,
      },
      styles.badge,
      badge.style,
    ];

    content.push(<Badge key="badge" {...badge} style={badgeStyles} onLayout={onBadgeLayout} />);
  }

  return (
    <View {...props} style={rootStyles}>
      {content}
    </View>
  );
};

const createStyleSheet = ({ Colors, Fonts }: Theme) =>
  StyleSheet.create({
    root: {
      alignSelf: 'center',
      alignItems: 'center',
      justifyContent: 'center',
      backgroundColor: Colors.gray2,
    },
    image: {
      width: '100%',
      height: '100%',
    },
    text: {
      ...Fonts.bold,
      color: Colors.white,
    },
    badge: {
      zIndex: 1,
      position: 'absolute',
      shadowColor: Colors.black,
      shadowOffset: {
        width: 0,
        height: 1,
      },
      shadowOpacity: 0.2,
      shadowRadius: 1,
      elevation: 1,
    },
  });

export type Props = AvatarProps & WithThemeProps;

export default withTheme(Avatar, createStyleSheet, 'Avatar');
