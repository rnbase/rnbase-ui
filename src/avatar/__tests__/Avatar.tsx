import React from 'react';
import { Image } from 'react-native';

// Note: test renderer must be required after react-native.
import TestRenderer, { act } from 'react-test-renderer';

/**
 * Under test
 */
import Avatar, { Props } from '../Avatar';

const size = 50;

const createElement = (props: Props) => <Avatar {...props} />;

const createRenderer = (props: Props) => TestRenderer.create(createElement(props));

it('should render default image', () => {
  expect(createRenderer({ size })).toMatchSnapshot();
});

it('should render image', () => {
  expect(createRenderer({ size, image: { uri: 'image.png' } })).toMatchSnapshot();
});

it('should render image as square', () => {
  expect(createRenderer({ size, shape: 'square', image: { uri: 'image.png' } })).toMatchSnapshot();
});

it('should render gravatar', () => {
  expect(createRenderer({ size, email: 'user@email.com' })).toMatchSnapshot();
});

it('should render initials', () => {
  expect(createRenderer({ size, name: 'User Name' })).toMatchSnapshot();
});

it('should fallback to initials', () => {
  const tree = createRenderer({ size, name: 'User Name', email: 'user@email.com' });
  const image = tree.root.findByType(Image);

  act(() => image.props.onError());

  expect(tree).toMatchSnapshot();
});

it('should render default image if no image', () => {
  const tree = createRenderer({
    size,
    image: { uri: 'image.png' },
    defaultImage: { uri: 'default.png' },
  });
  const image = tree.root.findByType(Image);

  act(() => image.props.onError());

  expect(tree).toMatchSnapshot();
});

it('should render default image if no gravatar', () => {
  const tree = createRenderer({
    size,
    email: 'user@email.com',
    defaultImage: { uri: 'default.png' },
  });
  const image = tree.root.findByType(Image);

  act(() => image.props.onError());

  expect(tree).toMatchSnapshot();
});

it('should render default image if empty name', () => {
  const tree = createRenderer({ size, name: '', defaultImage: { uri: 'default.png' } });
  const image = tree.root.findByType(Image);

  act(() => image.props.onError());

  expect(tree).toMatchSnapshot();
});
