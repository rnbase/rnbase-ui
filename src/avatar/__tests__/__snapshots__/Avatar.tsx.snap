// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fallback to initials 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 25,
        "height": 50,
        "width": 50,
      },
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#aeaeb2",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "fontSize": 20,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#ffffff",
          "fontFamily": "System",
          "fontWeight": "700",
          "includeFontPadding": false,
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    UN
  </Text>
</View>
`;

exports[`should render as circle with bottom-left badge 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 25,
        "height": 50,
        "width": 50,
      },
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#aeaeb2",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLayout={[Function]}
    source={
      Object {
        "testUri": "../../../src/avatar/assets/default.png",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 25,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
  <Badge
    onLayout={[Function]}
    style={
      Array [
        Object {
          "bottom": -2.5,
          "left": -2.5,
        },
        Object {
          "elevation": 1,
          "position": "absolute",
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 1,
          "zIndex": 1,
        },
        undefined,
      ]
    }
    value={1}
  />
</View>
`;

exports[`should render as circle with bottom-right badge 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 25,
        "height": 50,
        "width": 50,
      },
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#aeaeb2",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLayout={[Function]}
    source={
      Object {
        "testUri": "../../../src/avatar/assets/default.png",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 25,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
  <Badge
    onLayout={[Function]}
    style={
      Array [
        Object {
          "bottom": -2.5,
          "right": -2.5,
        },
        Object {
          "elevation": 1,
          "position": "absolute",
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 1,
          "zIndex": 1,
        },
        undefined,
      ]
    }
    value={1}
  />
</View>
`;

exports[`should render as circle with top-left badge 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 25,
        "height": 50,
        "width": 50,
      },
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#aeaeb2",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLayout={[Function]}
    source={
      Object {
        "testUri": "../../../src/avatar/assets/default.png",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 25,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
  <Badge
    onLayout={[Function]}
    style={
      Array [
        Object {
          "left": -2.5,
          "top": -2.5,
        },
        Object {
          "elevation": 1,
          "position": "absolute",
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 1,
          "zIndex": 1,
        },
        undefined,
      ]
    }
    value={1}
  />
</View>
`;

exports[`should render as circle with top-right badge 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 25,
        "height": 50,
        "width": 50,
      },
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#aeaeb2",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLayout={[Function]}
    source={
      Object {
        "testUri": "../../../src/avatar/assets/default.png",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 25,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
  <Badge
    onLayout={[Function]}
    style={
      Array [
        Object {
          "right": -2.5,
          "top": -2.5,
        },
        Object {
          "elevation": 1,
          "position": "absolute",
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 1,
          "zIndex": 1,
        },
        undefined,
      ]
    }
    value={1}
  />
</View>
`;

exports[`should render as rounded square 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 10,
        "height": 50,
        "width": 50,
      },
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#aeaeb2",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLayout={[Function]}
    source={
      Object {
        "testUri": "../../../src/avatar/assets/default.png",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 5,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`should render as square with bottom-left badge 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 0,
        "height": 50,
        "width": 50,
      },
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#aeaeb2",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLayout={[Function]}
    source={
      Object {
        "testUri": "../../../src/avatar/assets/default.png",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 0,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
  <Badge
    onLayout={[Function]}
    style={
      Array [
        Object {
          "bottom": -5,
          "left": -5,
        },
        Object {
          "elevation": 1,
          "position": "absolute",
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 1,
          "zIndex": 1,
        },
        undefined,
      ]
    }
    value={1}
  />
</View>
`;

exports[`should render as square with bottom-right badge 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 0,
        "height": 50,
        "width": 50,
      },
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#aeaeb2",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLayout={[Function]}
    source={
      Object {
        "testUri": "../../../src/avatar/assets/default.png",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 0,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
  <Badge
    onLayout={[Function]}
    style={
      Array [
        Object {
          "bottom": -5,
          "right": -5,
        },
        Object {
          "elevation": 1,
          "position": "absolute",
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 1,
          "zIndex": 1,
        },
        undefined,
      ]
    }
    value={1}
  />
</View>
`;

exports[`should render as square with top-left badge 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 0,
        "height": 50,
        "width": 50,
      },
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#aeaeb2",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLayout={[Function]}
    source={
      Object {
        "testUri": "../../../src/avatar/assets/default.png",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 0,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
  <Badge
    onLayout={[Function]}
    style={
      Array [
        Object {
          "left": -5,
          "top": -5,
        },
        Object {
          "elevation": 1,
          "position": "absolute",
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 1,
          "zIndex": 1,
        },
        undefined,
      ]
    }
    value={1}
  />
</View>
`;

exports[`should render as square with top-right badge 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 0,
        "height": 50,
        "width": 50,
      },
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#aeaeb2",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLayout={[Function]}
    source={
      Object {
        "testUri": "../../../src/avatar/assets/default.png",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 0,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
  <Badge
    onLayout={[Function]}
    style={
      Array [
        Object {
          "right": -5,
          "top": -5,
        },
        Object {
          "elevation": 1,
          "position": "absolute",
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 1,
          "zIndex": 1,
        },
        undefined,
      ]
    }
    value={1}
  />
</View>
`;

exports[`should render default image 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 25,
        "height": 50,
        "width": 50,
      },
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#aeaeb2",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLayout={[Function]}
    source={
      Object {
        "testUri": "../../../src/avatar/assets/default.png",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 25,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`should render default image if empty name 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 25,
        "height": 50,
        "width": 50,
      },
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#aeaeb2",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLayout={[Function]}
    source={
      Object {
        "uri": "default.png",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 25,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`should render default image if no gravatar 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 25,
        "height": 50,
        "width": 50,
      },
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#aeaeb2",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLayout={[Function]}
    source={
      Object {
        "uri": "default.png",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 25,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`should render default image if no image 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 25,
        "height": 50,
        "width": 50,
      },
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#aeaeb2",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLayout={[Function]}
    source={
      Object {
        "uri": "default.png",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 25,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`should render gravatar 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 25,
        "height": 50,
        "width": 50,
      },
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#aeaeb2",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLayout={[Function]}
    source={
      Object {
        "uri": "https://www.gravatar.com/avatar/b58c6f14d292556214bd64909bcdb118?s=100&d=404",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 25,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`should render image 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 25,
        "height": 50,
        "width": 50,
      },
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#aeaeb2",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLayout={[Function]}
    source={
      Object {
        "uri": "image.png",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 25,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`should render image as square 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 0,
        "height": 50,
        "width": 50,
      },
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#aeaeb2",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLayout={[Function]}
    source={
      Object {
        "uri": "image.png",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 0,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`should render initials 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 25,
        "height": 50,
        "width": 50,
      },
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#aeaeb2",
        "justifyContent": "center",
      },
      undefined,
      Object {
        "backgroundColor": "hsl(-216, 75%, 50%)",
      },
    ]
  }
>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "fontSize": 20,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#ffffff",
          "fontFamily": "System",
          "fontWeight": "700",
          "includeFontPadding": false,
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    UN
  </Text>
</View>
`;
